File: agent/tasks/sidechain.py
from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Any


class SidechainTranscript:
    def __init__(self, root: Path | str, task_id: str) -> None:
        self.root = Path(root)
        self.task_id = task_id
        self.path = self.root / "memory" / "tasks" / task_id / "transcript.jsonl"

    def append(self, message: dict[str, Any]) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        payload = dict(message)
        payload.setdefault("task_id", self.task_id)
        payload.setdefault("sidechain", True)
        payload.setdefault("ts", time.time())
        with self.path.open("a", encoding="utf-8") as file:
            file.write(json.dumps(payload, ensure_ascii=False) + "\n")

    def extend(self, messages: list[dict[str, Any]]) -> None:
        for message in messages:
            self.append(message)

    def read(self, *, offset: int = 0, limit: int = 100) -> dict[str, Any]:
        messages: list[dict[str, Any]] = []
        next_offset = 0
        if not self.path.exists():
            return {"messages": [], "nextOffset": 0, "path": str(self.path)}
        with self.path.open("r", encoding="utf-8") as file:
            for line_number, line in enumerate(file):
                next_offset = line_number + 1
                if line_number < offset:
                    continue
                if len(messages) >= limit:
                    break
                try:
                    payload = json.loads(line)
                except json.JSONDecodeError:
                    continue
                if isinstance(payload, dict):
                    messages.append(payload)
        return {
            "messages": messages,
            "nextOffset": min(next_offset, offset + len(messages)),
            "path": str(self.path),
        }

File: packages/core/src/tasks/sidechain.ts
import {
  appendFileSync,
  closeSync,
  constants,
  existsSync,
  lstatSync,
  mkdirSync,
  openSync,
  readFileSync,
  realpathSync,
} from 'node:fs'
import { dirname, isAbsolute, relative, resolve, sep } from 'node:path'

export class SidechainTranscript {
  readonly root: string
  readonly taskId: string
  readonly path: string
  private readonly tasksRoot: string
  private readonly taskRoot: string

  constructor(root: string, taskId: string) {
    this.root = resolve(root)
    this.taskId = safeTaskId(taskId)
    this.tasksRoot = resolve(this.root, 'tasks')
    this.taskRoot = resolve(this.tasksRoot, this.taskId)
    assertContainedPath(this.tasksRoot, this.taskRoot)
    this.path = resolve(this.taskRoot, 'transcript.jsonl')
    assertContainedPath(this.tasksRoot, this.path)
  }

  append(message: Record<string, unknown>): void {
    mkdirSync(dirname(this.path), { recursive: true })
    this.assertExistingPathBoundary()
    const payload = {
      ...message,
      task_id: message.task_id ?? this.taskId,
      sidechain: message.sidechain ?? true,
      ts: message.ts ?? Date.now() / 1000,
    }
    const descriptor = openSync(
      this.path,
      constants.O_APPEND |
        constants.O_CREAT |
        constants.O_WRONLY |
        noFollowFlag(),
      0o600,
    )
    try {
      appendFileSync(descriptor, JSON.stringify(payload) + '\n', 'utf8')
    } finally {
      closeSync(descriptor)
    }
  }

  extend(messages: Array<Record<string, unknown>>): void {
    for (const message of messages) this.append(message)
  }

  read(opts: { offset?: number; limit?: number } = {}): {
    messages: Array<Record<string, any>>
    nextOffset: number
    path: string
  } {
    const offset = Math.max(0, Math.trunc(opts.offset ?? 0))
    const limit = Math.max(0, Math.trunc(opts.limit ?? 100))
    const messages: Array<Record<string, any>> = []
    let nextOffset = 0
    if (!existsSync(this.path))
      return { messages: [], nextOffset: 0, path: this.path }
    this.assertExistingPathBoundary()
    const descriptor = openSync(this.path, constants.O_RDONLY | noFollowFlag())
    let text = ''
    try {
      text = readFileSync(descriptor, 'utf8')
    } finally {
      closeSync(descriptor)
    }
    const lines = text.split('\n')
    for (let lineNumber = 0; lineNumber < lines.length; lineNumber++) {
      const line = lines[lineNumber]
      if (!line) continue
      nextOffset = lineNumber + 1
      if (lineNumber < offset) continue
      if (messages.length >= limit) break
      try {
        const payload = JSON.parse(line)
        if (payload && typeof payload === 'object' && !Array.isArray(payload))
          messages.push(payload)
      } catch {
        continue
      }
    }
    return {
      messages,
      nextOffset: Math.min(nextOffset, offset + messages.length),
      path: this.path,
    }
  }

  private assertExistingPathBoundary(): void {
    for (const path of [this.tasksRoot, this.taskRoot, this.path]) {
      if (existsSync(path) && lstatSync(path).isSymbolicLink())
        throw new Error('sidechain transcript path contains a symlink')
    }
    if (!existsSync(this.tasksRoot) || !existsSync(this.taskRoot)) return
    const tasksReal = realpathSync(this.tasksRoot)
    const taskReal = realpathSync(this.taskRoot)
    assertContainedPath(tasksReal, taskReal)
    if (existsSync(this.path))
      assertContainedPath(tasksReal, realpathSync(this.path))
  }
}

function safeTaskId(value: unknown): string {
  const taskId = String(value ?? '').trim()
  if (!/^[A-Za-z0-9_-][A-Za-z0-9_.:-]*$/.test(taskId) || taskId.includes('..'))
    throw new Error('invalid task id')
  return taskId
}

function assertContainedPath(parent: string, candidate: string): void {
  const rel = relative(parent, candidate)
  if (rel === '..' || rel.startsWith(`..${sep}`) || isAbsolute(rel))
    throw new Error('sidechain transcript is outside task directory')
}

function noFollowFlag(): number {
  return typeof constants.O_NOFOLLOW === 'number' ? constants.O_NOFOLLOW : 0
}
